### PR TITLE
add service context to datafetchers

### PR DESCRIPTION
**Pull request by bhavinshah7**
# What Changed
Add ServiceContext interface to datafetchers

# Why
ServiceContext will be used by gateway to analyze metrics per subgraph/service provider. 

Todo:

- [ ] Add tests
- [ ] Add docs

### CODE REVIEW

File: src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java [hunk 1]
```diff
@@ -7,16 +7,18 @@
 import lombok.Getter;
 
 @Getter
-public class ServiceDataFetcher implements DataFetcher {
+public class ServiceDataFetcher implements DataFetcher, ServiceContext {
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java:10]: I am curious about this design pattern, given the fact that the serviceMetadata that has serviceProvider that has namespace

Comment (ashpak-shaikh) [reply src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java:10]: We can possibly do 
```
NamespaceAwareDataFetcher implements DataFetcher {
...
}
```

and all DF extend this one instead DataFetcher directly. you can also do ServiceAwareDF if all DF has that information

Comment (bhavinshah7) [reply src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java:10]: That is a good idea, i will update.

File: src/main/java/com/intuit/graphql/orchestrator/stitching/XtextStitcher.java [hunk 2]
```diff
@@ -247,8 +247,8 @@ private GraphQLCodeRegistry.Builder getCodeRegistry(UnifiedXtextGraph mergedGrap
 
         builder.dataFetcher(coordinates,
             dataFetcherContext.getServiceType() == ServiceType.REST
-                ? new RestDataFetcher(serviceMetadata)
-                : new ServiceDataFetcher(serviceMetadata)
+                ? new RestDataFetcher(serviceMetadata, dataFetcherContext.getNamespace())
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/stitching/XtextStitcher.java:250]: Why do we need to pass namespace explicitly? metadata should already have namespace

File: src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java [hunk 3]
```diff
@@ -7,16 +7,18 @@
...
    * One of Query or Mutation or Subscription
    */
   private final ServiceMetadata serviceMetadata;
+  private final String namespace;
 
 
-  public ServiceDataFetcher(ServiceMetadata serviceMetadata) {
+  public ServiceDataFetcher(ServiceMetadata serviceMetadata, String namespace) {
     this.serviceMetadata = serviceMetadata;
+    this.namespace = namespace;
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java:21]: this.namespace = serviceMetadata.serviceProvider.namespace ?

**Review by kmoore-intuit — APPROVED** [review]

File: src/main/java/com/intuit/graphql/orchestrator/datafetcher/FieldResolverDirectiveDataFetcher.java [hunk 4]
```diff
@@ -6,20 +6,26 @@
 import graphql.schema.DataFetcher;
 import graphql.schema.DataFetchingEnvironment;
 import java.util.Objects;
+import lombok.Getter;
 
-public class FieldResolverDirectiveDataFetcher implements DataFetcher<Object> {
+@Getter
+public class FieldResolverDirectiveDataFetcher implements ServiceAwareDataFetcher<Object> {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/datafetcher/FieldResolverDirectiveDataFetcher.java:12]: I'm thinking of standardizing to `... implement `ServiceAwareDataFetcher<CompletableFuture<DataFetcherResult<Object>>>`.

FieldResolverBatrchLoader.java has this signature

```
public CompletionStage<List<DataFetcherResult<Object>>> load(final List<DataFetchingEnvironment> dataFetchingEnvironments)
```

File: src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceAwareDataFetcher.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.intuit.graphql.orchestrator.datafetcher;
+
+import graphql.schema.DataFetcher;
+
+public interface ServiceAwareDataFetcher<T> extends DataFetcher<T> {
+  String getNamespace();
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceAwareDataFetcher.java:6]: Also making serviceMedata accessible would help get access to otherinfo

File: src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java [hunk 6]
```diff
@@ -2,12 +2,11 @@
 
 import com.intuit.graphql.orchestrator.schema.ServiceMetadata;
 import graphql.GraphQLContext;
-import graphql.schema.DataFetcher;
 import graphql.schema.DataFetchingEnvironment;
 import lombok.Getter;
 
 @Getter
-public class ServiceDataFetcher implements DataFetcher {
+public class ServiceDataFetcher implements ServiceAwareDataFetcher {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/datafetcher/ServiceDataFetcher.java:9]: same... if possible, would be good to have `ServiceAwareDataFetcher<CompletableFuture<DataFetcherResult>>`

**Pull request merged**